lib/tts_isolate.dart

```diff
@@ -53,41 +54,35 @@ void onStart(ServiceInstance service) async {
   WidgetsFlutterBinding.ensureInitialized();
   await Firebase.initializeApp();
 
-  if (service is AndroidServiceInstance) {
-    service.on('setAsForeground').listen((event) async {
-      service.setAsForegroundService();
-      // Setting up the notification at the very beginning
-    });
-
-    service.on('setAsBackground').listen((event) {
-      service.setAsBackgroundService();
-    });
-  }
-
-  // Handle other service events
-  service.on('stopService').listen((event) {
-    // print("Stopping this service right now");
-    TextToSpeechPlugin.stopSpeaking();
-  });
-
-  service.on('startTts').listen((event) async {
-    await Isolate.spawn(isolateMain, ReceivePort().sendPort);
-  });
-
-  final prefs = await StreamingSharedPreferences.instance;
-  prefs.getString('tts_channel', defaultValue: '{}').switchMap((channel) {
-    if (channel.isNotEmpty && channel != "{}") {
-      return FirebaseFirestore.instance
-          .collection('channels')
-          .where('channelId', isEqualTo: channel)
-          .snapshots();
-    } else {
-      return const Stream.empty();
-    }
-  }).listen((snapshot) {
-    for (final change in snapshot.docChanges) {
-      if (change.type == DocumentChangeType.added) {
-        vocalizeMessage(change.doc.data());
+  channelSubscription = service.on("setTtsChannel").listen((event) async {
+    if (event != null) {
+      Map<String, dynamic>? args = event;
+      if (args['channel'] == null) {
+        await ttsQueue.clear();
+        messagesSubscription?.cancel();
+      } else {
+        var channelData = args["channel"];
+        var ttsCurrentChannel =
+            "${channelData['provider']}:${channelData['channelId']}";
+        messagesSubscription?.cancel();
+        messagesSubscription = FirebaseFirestore.instance
+            .collection('channels')
+            .doc(ttsCurrentChannel)
+            .collection('messages')
+            .orderBy('timestamp', descending: true)
+            .limit(1)
+            .snapshots()
+            .listen((latestMessage) async {
```

Comment:
I think there's a subtle bug here that this will read the last message whereas we want to only read future messages. The way around this is to keep a timestamp when the listener is added and reject messages that are added before that timestamp.

Reply:
This should resolve it